### PR TITLE
Enumerables

### PR DESCRIPTION
**Pull request by thedekerone**
In this pull request:
- I finished building all the enumerable methods.
- Tested all the methods
- Fixed Linter errors

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -17,6 +17,9 @@ Metrics/ClassLength:
 Metrics/BlockLength:
   ExcludedMethods: ['describe']
   Max: 30
+Metrics/ModuleLength:
+  ExcludedMethods: ['describe']
+  Max: 200
```
Comment (shshamim63) [inline .rubocop.yml]: You should not overwrite the default config file provided by Microverse. Instead, you can use the inline rule disable approach to get rid of an error that you think going to take to many tasks to resolve it. But if you think that a ruleset is important and it should be added check out the instruction in this [link](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml)

File: README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Methods project
```
Comment (shshamim63) [inline README.md]: I think you can rename it into a more relational name. Something like. `Enumerable Methods Project`

File: README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Methods project
+
+This is a project of two algorithms about Enumerble methods in ruby.
```
Comment (shshamim63) [inline README.md]: Can you rephrase this sentence? It does not seem clear to me.

File: my_each.rb [hunk 4]
```diff
@@ -1,6 +1,168 @@
...
+        result = compare.class.eql?(Class) ? element.is_a?(compare) : element.eql?(compare)
+      elsif block_given?
+        result = yield element
+
+      end
+      break unless result
+    end
+
+    result
+  end
```
Comment (shshamim63) [inline my_each.rb]: `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- when no block or argument is given, returns true when none of the collection members are false or nil
```ruby
[1, 2, 3].my_all? #should return true
[1, 2, nil].my_all? #should return false
[1, false, nil].my_all? #should return false
```
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
%w[dog door rod blade].my_all?(/d/)   #should return true.
%w[dog door rod blade].my_all?(/t/)   #should return false.
```

File: my_each.rb [hunk 5]
```diff
@@ -1,6 +1,168 @@
...
+    my_each do |element|
+      if compare
+        result = compare.class.eql?(Class) ? element.is_a?(compare) : element.eql?(compare)
+      elsif block_given?
+        result = yield element
+      end
+      break if result
+    end
+    result
+  end
```
Comment (shshamim63) [inline my_each.rb]: `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
[1, 2, 3].my_any? #should return true
[1, 2, nil].my_any? #should return true
[false, false, nil].my_any? #should return false
```
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_any?(/t/)    #should return false.
%w[dog door rod blade].my_any?(/d/)   #should return true.
```

File: my_each.rb [hunk 6]
```diff
@@ -1,6 +1,168 @@
...
+                 compare.class.eql?(Class) ? !element.is_a?(compare) : !element.eql?(compare)
+               elsif block_given?
+                 !(yield element)
+               else
+                 !element
+               end
+      break unless result
+    end
+    result
+  end
```
Comment (shshamim63) [inline my_each.rb]: `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/)   #should return true
%w[dog door rod blade].my_none?(/d/)   #should return false
```

File: my_each.rb [hunk 7]
```diff
@@ -1,6 +1,168 @@
...
+    my_each do |element|
+      param = if arg
+                arg
+              elsif block_given?
+                yield element
+              end
+      counter += 1 if param.eql?(true) || param.eql?(element)
+    end
+    counter
+  end
```
Comment (shshamim63) [inline my_each.rb]: `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns the number of items in enum through enumeration
```ruby
[1, 2, 3].my_count  #should return 3
```
- counts the number of items in enum that are equal to item if an argument is given
```ruby
[1, 1, 2 ,3].my_count(1)  #should return 2
```
- counts the number of elements yielding a true value if a block is given
```ruby
[1, 2, 3].my_count {|element| element > 1}    #should return 2
```

File: my_each.rb [hunk 8]
```diff
@@ -1,6 +1,168 @@
...
+    end
+    initial_value = 0 if use == :+
+
+    my_each do |element|
+      result = use&.to_proc&.call(initial_value, element)
+      result = yield initial_value, element if block_given?
+      initial_value = result
+    end
+    initial_value
+  end
```
Comment (shshamim63) [inline my_each.rb]: `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```

File: my_each.rb [hunk 9]
```diff
@@ -1,6 +1,168 @@
...
+end
+
+def multiply_els
+  my_inject(1) do |acumulator, element|
+    acumulator * element
+  end
+end
```
Comment (shshamim63) [inline my_each.rb]: multiply_els should accept a parameter. I can not see it. The methods should look something like
```ruby
def multiply_els array
  array.my_inject(1) do |acumulator, element|
    acumulator * element
  end
end

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :tractor: 
___
Hi @thedekerone :wave: ,
Awesome job :clap:  almost near to complete this project. Let's fix some issues and declare it as complete.
___
:rocket: I saw you have created a new branch to get a review after the first one. Actually you have to stick to one PR(Pull Request) to complete a project/milestone. The previous reviewer asked you to rename the branch, actually, she was referring to change the title of the PR. The title of the PR should be self-descriptive. So please stay with this PR for further review.
___
### :spades: Mandatory Changes :spades: 
___
1. The following methods are not covering some edge cases. Please check the comments added to the code blocks.
- my_all?
- my_any?
- my_none?
- my_count
- my_inject
2. The multiply_els method is not fully configured yet. Check the comment attached to it.
3. You can brush the readme a little bit more to make it more informative.
4. You should not alter the basic configuration of the 'rubocop'.
image
___
#### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Comment by thedekerone** [comment]
Returned .rubocop file to previews state and made all the changes.

**Comment by Anna-Myzukina** [comment]
# Status: [changes required]


### Good job with the project. Few steps and your project can be approved:
`Implementation details`


* Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history from [Pathwright]()
screenshot 


#### Please check the ruby doc for these [methods](https://ruby-doc.org/core-2.6.5/Enumerable.html) behaviour. Methods to look out for are the following:


* my_all?
* my_none?
* my_inject

### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*
*Happy coding * <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

File: my_each.rb [hunk 10]
```diff
@@ -1,6 +1,141 @@
...
+                 yield element
+
+               else
+                 element ^ 1
+               end
+      break unless result
+    end
+
+    result
+  end
```
Comment (stratospherique) [inline my_each.rb]: * [ ] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_all? when no block or argument is given returns true when none of the collection members are false or nil
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all? == array.all? # false
  ```
* [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all?(5) == array.all?(5) # false
  ```

File: my_each.rb [hunk 11]
```diff
@@ -1,6 +1,141 @@
...
+                 compare.class.eql?(Class) ? element.is_a?(compare) : element.to_s.match?(compare.to_s)
+               elsif block_given?
+                 yield element
+               else
+                 element ^ 1
+               end
+      break if result
+    end
+    result
+  end
```
Comment (stratospherique) [inline my_each.rb]: * [ ] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
  ```ruby
    array = [nil, false, true, []]
    p array.my_any? == array.any? # false
  ```

File: my_each.rb [hunk 12]
```diff
@@ -1,6 +1,141 @@
...
+                 compare.class.eql?(Class) ? !element.is_a?(compare) : !element.to_s.match?(compare)
+               elsif block_given?
+                 !(yield element)
+               else
+                 !element
+               end
+      break unless result
+    end
+    result
+  end
```
Comment (stratospherique) [inline my_each.rb]: * [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_none?(5) == array.none?(5) # false
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested changes :clap: However, some changes need to be implemented in order for us to approve you project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: my_each.rb (hunk 11 above)
Comment (thedekerone) [reply my_each.rb]: I changed my_all and my_none, but this method was already working

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :tada: :clap: 
Hi there @thedekerone :hand: 
All required changes were implemented. Congratulations on the hard work :+1: 
You can merge the branch now
Good luck with the next project.

Happy coding :computer: 
**Sergio Torres**
